### PR TITLE
revise apiservice webhook

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 
https://github.com/pingcap/tidb-operator/issues/1317
https://github.com/pingcap/tidb-operator/issues/1316
https://github.com/pingcap/tidb-operator/issues/1315

This request would fix the problems left in https://github.com/pingcap/tidb-operator/pull/1288

### What is changed and how does it work?

1.  Revising `manifests/patch-e2e.sh`. As cabundle can only be accepted as null by kube-apiserver with version >= 1.13.0, now `patch-e2e.sh` would only patch cabundle if kube-apiserver with version < 1.13.0 in e2e test

2. Using one single api path for all webhook admission path.

3. Removing all the `Lister` and using kubecli instead.

### CODE REVIEW

File: manifests/patch-e2e.sh [hunk 1]
```diff
@@ -38,20 +32,32 @@ namespace=${namespace:-tidb-admin}
...
+esac
 
+if [ "$bundle" = true ] ; then
+
+CA_BUNDLE=$(kubectl get configmap -n kube-system extension-apiserver-authentication -o=jsonpath='{.data.client-ca-file}' | base64 | tr -d '\n')
 
 case ${os} in
   linux)
     sed -i "s/caBundle: .*$/caBundle: ${CA_BUNDLE}/g" $CURDIR/webhook.yaml
-    sed -i "s/namespace: \${NAMESPACE}/namespace: ${namespace}/g" $CURDIR/webhook.yaml
```
Comment (weekface) [inline manifests/patch-e2e.sh]: why move this line to 37?

**Comment by weekface** [comment]
In webhook, there are many actions are deleting something, for example: `pdDeleter`, `tikvDeleter`, so for safe, we should not use any `xxxLister` cache in the webhook.

File: manifests/patch-e2e.sh (hunk 1 above)
Comment (Yisaer) [reply manifests/patch-e2e.sh]: There are 2 `sed` actions in this script. The first `sed` would change the namespace for `webhook.yaml`. And the second `sed` would change the `cabundle: null` into `cabundle: <cabundle token>` depends on kube-apiserver version.

**Comment by Yisaer** [comment]
Update, now all the lister is removed.

File: cmd/admission-webhook/main.go [hunk 2]
```diff
@@ -51,10 +56,32 @@ func main() {
...
 	}
-	stsAh := &webhook.StatefulSetAdmissionHook{}
-	cmd.RunAdmissionServer(podAh, stsAh)
+	run(flagset, ah)
+}
+
+func run(flagset *flag.FlagSet, admissionHooks ...apiserver.AdmissionHook) {
+
+	if len(os.Getenv("GOMAXPROCS")) == 0 {
+		runtime.GOMAXPROCS(runtime.NumCPU())
```
Comment (aylei) [inline cmd/admission-webhook/main.go]: why is this necessary?

File: pkg/webhook/admission_hooks.go [hunk 3]
```diff
@@ -19,60 +19,68 @@ import (
...
+		return a.stsAC.AdmitStatefulSets(ar)
+	default:
+		return a.unknownAdmissionRequest(ar)
 	}
-	return a.podAC.AdmitPods(ar)
 }
 
 // any special initialization goes here
-func (a *PodAdmissionHook) Initialize(cfg *rest.Config, stopCh <-chan struct{}) error {
+func (a *AdmissionHook) Initialize(cfg *rest.Config, stopCh <-chan struct{}) error {
```
Comment (aylei) [inline pkg/webhook/admission_hooks.go]: shall we add

```
if a.initialized {
  return nil
}
```
?

Comment (Yisaer) [reply pkg/webhook/admission_hooks.go]: good advice, updated.

File: cmd/admission-webhook/main.go [hunk 4]
```diff
@@ -15,29 +15,34 @@ package main
 
 import (
 	"flag"
+	"github.com/openshift/generic-admission-server/pkg/apiserver"
```
Comment (tennix) [inline cmd/admission-webhook/main.go]: Could you configure your editor to use `goimports` to format code? This will ensure the `import` package order consistent.

File: cmd/admission-webhook/main.go (hunk 2 above)
Comment (tennix) [reply cmd/admission-webhook/main.go]: +1. If we need to configure GOMAXPROCS, we can just use uber's `automaxprocs` package.

Comment (Yisaer) [reply cmd/admission-webhook/main.go]: This is introduced in the orgin framework by [cmd.RunAdmissionServer](https://github.com/openshift/generic-admission-server/blob/92cbaa8d51b1b9fa33a7eda76683b3482590e534/pkg/cmd/cmd.go#L28).
Because I want to import my extra flagset, I copied this function like [quack](https://github.com/pusher/quack/blob/5d7531613d6bf969b9018fde40a5cffb7d3efd7e/cmd/quack/main.go#L34).

File: cmd/admission-webhook/main.go (hunk 4 above)
Comment (Yisaer) [reply cmd/admission-webhook/main.go]: updated and add `goimport` into FileWatcher. Missing it was a mistake.

File: cmd/admission-webhook/main.go (hunk 2 above)
Comment (tennix) [reply cmd/admission-webhook/main.go]: I think Go runtime will use NumCPU if the GOMAXPROCS=0 automatically.

File: cmd/admission-webhook/main.go [hunk 5]
```diff
@@ -51,10 +57,32 @@ func main() {
...
+	stopCh := genericapiserver.SetupSignalHandler()
+
+	cmd := server.NewCommandStartAdmissionServer(os.Stdout, os.Stderr, stopCh, admissionHooks...)
+
+	// Add admission hook flags
+	cmd.Flags().AddGoFlagSet(flagset)
+
+	// Flags for glog
+	cmd.PersistentFlags().AddGoFlagSet(flag.CommandLine)
+	// Fix glog printing "Error: logging before flag.Parse"
```
Comment (tennix) [inline cmd/admission-webhook/main.go]: Does the klog still have this issue?

File: cmd/admission-webhook/main.go (hunk 2 above)
Comment (Yisaer) [reply cmd/admission-webhook/main.go]: updated, removing this logic.

File: cmd/admission-webhook/main.go [hunk 6]
```diff
@@ -16,28 +16,34 @@ package main
...
 
 var (
+	flagset                  *flag.FlagSet
 	printVersion             bool
 	extraServiceAccounts     string
 	evictRegionLeaderTimeout time.Duration
 )
 
 func init() {
+	flagset = flag.NewFlagSet("tidb-admission-webhook", flag.ExitOnError)
```
Comment (tennix) [inline cmd/admission-webhook/main.go]: Do we need to introduce the flagset? If so, I think we should use only one of flagset or the standard library flag.

File: cmd/admission-webhook/main.go (hunk 5 above)
Comment (Yisaer) [reply cmd/admission-webhook/main.go]: updated, this logic was useless after testing and removed it now.

File: cmd/admission-webhook/main.go (hunk 6 above)
Comment (Yisaer) [reply cmd/admission-webhook/main.go]: updated, this `flagset` is not necessary and removed now.

File: manifests/patch-e2e.sh [hunk 7]
```diff
@@ -38,20 +32,32 @@ namespace=${namespace:-tidb-admin}
...
+    sed -i "" "s/namespace: \${NAMESPACE}/namespace: ${namespace}/g" $CURDIR/webhook.yaml
+    ;;
+    *)
+    echo "invalid os ${os}, only support Linux and Darwin" >&2
+    ;;
+esac
 
+if [ "$bundle" = true ] ; then
+
+CA_BUNDLE=$(kubectl get configmap -n kube-system extension-apiserver-authentication -o=jsonpath='{.data.client-ca-file}' | base64 | tr -d '\n')
```
Comment (tennix) [inline manifests/patch-e2e.sh]: Indent?

File: manifests/patch-e2e.sh [hunk 8]
```diff
@@ -38,20 +32,32 @@ namespace=${namespace:-tidb-admin}
 CURDIR=$(cd $(dirname ${BASH_SOURCE[0]}); pwd )
 os=`uname -s| tr '[A-Z]' '[a-z]'`
 
-CA_BUNDLE=$(kubectl get configmap -n kube-system extension-apiserver-authentication -o=jsonpath='{.data.client-ca-file}' | base64 | tr -d '\n')
-echo $CA_BUNDLE
+case ${os} in
+  linux)
+    sed -i "s/namespace: \${NAMESPACE}/namespace: ${namespace}/g" $CURDIR/webhook.yaml
```
Comment (tennix) [inline manifests/patch-e2e.sh]: Will this modify the file directly? This file is checked into the git. So I think it's better to do the modification on the fly by using pipe.

Comment (tennix) [reply manifests/patch-e2e.sh]: Or can we do the patch after apply to k8s?

Comment (Yisaer) [reply manifests/patch-e2e.sh]: Currently, this would only modify the file in e2e test. I think this could be marked as an issue to find a more elegant way in a future request.

File: manifests/patch-e2e.sh (hunk 7 above)
Comment (Yisaer) [reply manifests/patch-e2e.sh]: updated

File: manifests/patch-e2e.sh (hunk 8 above)
Comment (Yisaer) [reply manifests/patch-e2e.sh]: issed in https://github.com/pingcap/tidb-operator/issues/1358
